src/storage/src/source/reclock.rs

```diff
@@ -8,413 +8,668 @@
-            // and should be in one place only.
-            self.write_upper = match compare_and_append_result {
-                Ok(()) => new_upper,
-                Err(Upper(actual_upper)) => actual_upper,
-            };
+        }
 
-            while PartialOrder::less_than(&self.read_progress, &self.write_upper) {
+        // If there no updates we might still want to bump the upper time if enough time has passed
+        let mint_timestamp = if updates.is_empty() {
+            self.next_mint_timestamp().now_or_never()
+        } else {
+            Some(self.next_mint_timestamp().await)
+        };
+
+        if let Some(next_ts) = mint_timestamp {
+            let new_upper = Antichain::from_elem(next_ts + 1);
+            loop {
+                let upper = self.upper.clone();
+                let new_upper = new_upper.clone();
+                let updates = updates.iter().map(|(pid, diff)| (((), pid), next_ts, diff));
                 match self
-                    .timestamp_bindings_listener
-                    .next()
+                    .write_handle
+                    .compare_and_append(updates, upper, new_upper)
                     .await
-                    .expect("ListenStream doesn't end")
                 {
-                    ListenEvent::Progress(progress) => {
-                        assert!(
-                            timely::PartialOrder::less_equal(&self.read_progress, &progress),
-                            "{:?} PARTIAL ORDER: {:?} {:?}",
-                            self.name,
-                            self.read_progress,
-                            progress
-                        );
-                        self.read_progress = progress;
-                    }
-                    ListenEvent::Updates(updates) => {
-                        for ((_, value), timestamp, diff) in updates {
-                            let partition = value.expect("Unable to decode partition id");
-                            self.persisted_timestamp_bindings
-                                .entry(partition)
-                                .or_insert_with(VecDeque::new)
-                                .push_back((timestamp, diff));
-                        }
-                    }
+                    Ok(Ok(Ok(()))) => break,
+                    Ok(Ok(Err(actual_upper))) => return Err(actual_upper),
+                    Ok(Err(invalid_use)) => panic!("{invalid_use:?}"),
+                    Err(_external_err) => continue,
                 }
             }
+            self.sync(&new_upper).await;
+        }
+        Ok(())
+    }
 
-            debug_assert!(self.validate_persisted_bindings());
+    /// Produces a new timestamp suitable for minting bindings. The returned value is guaranteed to
+    /// be beyond the current `upper` frontier and a multiple of `update_interval_ms`.
+    async fn next_mint_timestamp(&mut self) -> Timestamp {
```

Comment:
what do you think of adding a TODO about large amount of clock skew?

Reply:
I tried to figure out how this operator would even know that skew had happened and couldn't. Maybe worth talking about it on slack if you have some ideas